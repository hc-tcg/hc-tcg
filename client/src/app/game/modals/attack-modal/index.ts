export {default} from './attack-modal'
