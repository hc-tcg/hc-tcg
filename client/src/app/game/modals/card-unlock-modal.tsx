import CardList from 'components/card-list'
import Modal from 'components/modal'
import {getRollFail, getRollResult} from 'logic/permits/permits-selectors'
import {useDispatch, useSelector} from 'react-redux'
import css from './game-modals.module.scss'
import Button from 'components/button'
import {clearResult} from 'logic/permits/permits-actions'

function CardUnlockModal() {
	const dispatch = useDispatch()
	const onModalClose = () => {
		dispatch(clearResult())
	}

	const rollFail = useSelector(getRollFail)
	const rollResult = useSelector(getRollResult)

	return (
		<Modal
			closeModal={onModalClose}
			title={rollFail === '' ? 'New Permit Unlocked!' : 'Permit Purchase Problem!'}
		>
			<div className={css.description}>
				<div className={css.card}>
					{rollResult ? (
						<CardList cards={[{cardId: rollResult?.card, cardInstance: ''}]} />
					) : (
						rollFail
					)}
				</div>{' '}
			</div>
			<div className={css.options}>
				<Button onClick={onModalClose}>Ok</Button>
			</div>
		</Modal>
	)
}

export default CardUnlockModal
