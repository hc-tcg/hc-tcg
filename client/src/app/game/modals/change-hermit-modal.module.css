.description {
	padding: 10px;
}

.options {
	text-align: center;
	padding-bottom: 10px;
}

.options button + button {
	margin-left: 5px;
}

.turnEndNotification {
	font-size: 12px;
	margin: 0 auto;
	padding: 5px 10px 15px 10px;
}

.infoIcon {
	border: 2px solid black;
	width: 20px;
	height: 20px;
	border-radius: 10px;
	font-size: 16px;
	line-height: 16px;
	text-align: center;
	display: inline-block;
	vertical-align: baseline;
	margin-right: 5px;
	font-family: serif;
	font-weight: bold;
}
