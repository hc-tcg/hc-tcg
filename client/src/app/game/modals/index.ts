export {default as AttackModal} from './attack-modal'
export {default as ChangeHermitModal} from './change-hermit-modal'
export {default as ConfirmModal} from './confirm-modal'
export {default as EndTurnModal} from './end-turn-modal'
export {default as ForfeitModal} from './forfeit-modal'
export {default as UnmetConditionModal} from './unmet-condition-modal'
export {default as CopyAttackModal} from './copy-attack-modal'
export {default as SelectCardsModal} from './select-cards-modal'
