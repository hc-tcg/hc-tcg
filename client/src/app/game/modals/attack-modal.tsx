import Modal from 'components/modal'
import {useSelector, useDispatch} from 'react-redux'
import {CardInfoT, EffectCardT, HermitCardT} from 'types/cards'
import classnames from 'classnames'
import CARDS from 'server/cards'
import Strengths from 'server/const/strengths'
import {
	getPlayerActiveRow,
	getOpponentActiveRow,
	getMultiplier,
} from '../game-selectors'
import css from './attack-modal.module.css'

import {getPlayerId} from 'logic/session/session-selectors'
import {
	getAvailableActions,
	getPlayerStateById,
} from 'logic/game/game-selectors'
import {startAttack} from 'logic/game/game-actions'

const TYPED_CARDS = CARDS as Record<string, CardInfoT>

type Props = {
	closeModal: () => void
}
function AttackModal({closeModal}: Props) {
	// TODO - This whole file needs to be rafactored
	const dispatch = useDispatch()
	const activeRow = useSelector(getPlayerActiveRow)
	const opponentRow = useSelector(getOpponentActiveRow)
	const availableActions = useSelector(getAvailableActions)
	const playerId = useSelector(getPlayerId)
	const playerState = useSelector(getPlayerStateById(playerId))
	const singleUseCard = playerState?.board.singleUseCard
	const multiplier = useSelector(getMultiplier)

	if (!activeRow || !activeRow.hermitCard) return null
	if (!opponentRow || !opponentRow.hermitCard) return null

	const playerHermitInfo = TYPED_CARDS[
		activeRow.hermitCard.cardId
	] as HermitCardT
	const opponentHermitInfo = TYPED_CARDS[
		opponentRow.hermitCard.cardId
	] as HermitCardT
	const hermitFullName = playerHermitInfo.id.split('_')[0]

	// const playerEffectInfo = activeRow.effectCard
	// 	? (TYPED_CARDS[activeRow.effectCard.cardId] as EffectCardT)
	// 	: null
	const opponentEffectInfo = opponentRow.effectCard
		? (TYPED_CARDS[opponentRow.effectCard.cardId] as EffectCardT)
		: null
	const singleUseInfo = singleUseCard
		? (TYPED_CARDS[singleUseCard.cardId] as EffectCardT)
		: null

	const suAttackInfo =
		singleUseInfo && singleUseInfo.damage
			? {
					name: singleUseInfo.name,
					damage: singleUseInfo.damage.target || 0,
					afkDamage: singleUseInfo.damage.afkTarget || 0,
			  }
			: null

	const protectionAmount =
		suAttackInfo && singleUseInfo?.id === 'golden_axe'
			? 0
			: opponentEffectInfo?.protection?.target || 0

	const hasWeakness = Strengths[playerHermitInfo.hermitType].includes(
		opponentHermitInfo.hermitType
	)

	const handleAttack = (type: 'zero' | 'primary' | 'secondary') => {
		dispatch(startAttack(type))
		closeModal()
	}

	const effectAttack = () => handleAttack('zero')
	const primaryAttack = () => handleAttack('primary')
	const secondaryAttack = () => handleAttack('secondary')

	const renderAttack = (
		attackInfo: any,
		onClick: () => void,
		icon?: string
	) => {
		const baseDamage = icon ? 0 : attackInfo.damage
		const totalDamage = Math.max(
			baseDamage +
				(suAttackInfo ? suAttackInfo.damage : 0) +
				(hasWeakness && baseDamage > 0 ? 20 : 0) -
				(protectionAmount || 0),
			0
		)
		return (
			<div key={attackInfo.name} className={css.attack} onClick={onClick}>
				<div
					className={classnames(css.icon, {
						[css.effectIcon]: !!icon,
						[css.hermitIcon]: !icon,
					})}
				>
					<img src={icon || `/images/hermits-nobg/${hermitFullName}.png`} />
				</div>
				<div className={css.info}>
					<div className={css.name}>
						{attackInfo.name} -{' '}
						<span
							className={classnames(css.damage, {
								[css.specialMove]: !!attackInfo.power,
							})}
						>
							{totalDamage}
						</span>
						{multiplier ? (
							<span className={css.multiplier}> x{multiplier}</span>
						) : null}
					</div>
					<div className={css.description}>
						{icon ? null : (
							<>
								<div className={css.hermitDamage}>
									<img
										src={`/images/hermits-nobg/${hermitFullName}.png`}
										width="32"
									/>
								</div>
								<div className={css.damageAmount}>{attackInfo.damage}</div>
							</>
						)}
						{suAttackInfo ? (
							<>
								{!icon ? <div className={css.attackOperator}>+</div> : null}
								<img
									src={`/images/effects/${singleUseInfo?.id}.png`}
									width="16"
									height="16"
								/>
								<div className={css.damageAmount}>
									{suAttackInfo.damage}
									{suAttackInfo?.afkDamage ? (
										<>({suAttackInfo.afkDamage})</>
									) : null}
								</div>
							</>
						) : null}

						{hasWeakness && baseDamage > 0 ? (
							<>
								<div className={css.attackOperator}>+</div>
								<img src={`/images/weakness.png`} width="16" height="16" />
								<div className={css.damageAmount}>20</div>
							</>
						) : null}
						{opponentEffectInfo && protectionAmount ? (
							<>
								<div className={css.attackOperator}>-</div>
								<img
									src={`/images/effects/${opponentEffectInfo.id}.png`}
									width="16"
									height="16"
								/>
								<div className={css.damageAmount}>{protectionAmount}</div>
							</>
						) : null}
					</div>
				</div>
			</div>
		)
	}

	const attacks = []
	if (suAttackInfo && availableActions.includes('ZERO_ATTACK')) {
		attacks.push(
			renderAttack(
				suAttackInfo,
				effectAttack,
				`/images/effects/${singleUseInfo?.id}.png`
			)
		)
	}

	if (availableActions.includes('PRIMARY_ATTACK')) {
		attacks.push(renderAttack(playerHermitInfo.primary, primaryAttack))
	}

	if (availableActions.includes('SECONDARY_ATTACK')) {
		attacks.push(renderAttack(playerHermitInfo.secondary, secondaryAttack))
	}

	return (
		<Modal title="Attack" closeModal={closeModal}>
			<div className={css.attackModal}>
				{attacks.length ? (
					<>
						<div className={css.turnEndNotification}>
							<span className={css.infoIcon}>!</span>
							Attack is the last action of your turn.
						</div>
						<div className={css.turnEndNotification}>
							<span className={css.infoIcon}>i</span>
							Damage bonuses from special moves are NOT included in the preview.
						</div>
						{attacks}
					</>
				) : (
					<span className={css.noAttacks}>No attacks available.</span>
				)}
			</div>
		</Modal>
	)
}

export default AttackModal
