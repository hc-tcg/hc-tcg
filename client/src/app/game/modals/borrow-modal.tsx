import Modal from 'components/modal'
import {useDispatch, useSelector} from 'react-redux'
import css from './game-modals.module.scss'
import Button from 'components/button'
import {getPlayerState} from 'logic/game/game-selectors'
import {CARDS} from 'common/cards'

type Props = {
	closeModal: () => void
}
function BorrowModal({closeModal}: Props) {
	const dispatch = useDispatch()

	const handleAttach = () => {
		dispatch({type: 'BORROW_ATTACH'})
		closeModal()
	}

	const handleDiscard = () => {
		dispatch({type: 'BORROW_DISCARD'})
		closeModal()
	}

	const getBorrowedCard = () => {
		const playerState = useSelector(getPlayerState)
		if (!playerState) return null

		const custom = Object.values(playerState.custom)
		if (!custom.length) return null

		const borrowedId = custom[0].cardId
		const borrowedName = CARDS[borrowedId].name

		return borrowedName
	}

	return (
		<Modal title="Borrow" closeModal={handleDiscard}>
			<div className={css.description}>
				Would you like to attach or discard your opponents' {getBorrowedCard()} card?
			</div>
			<div className={css.options}>
				<Button onClick={handleAttach}>Attach</Button>
				<Button onClick={handleDiscard}>Discard</Button>
			</div>
		</Modal>
	)
}

export default BorrowModal
