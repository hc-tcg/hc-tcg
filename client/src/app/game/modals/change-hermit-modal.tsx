import Modal from 'components/modal'
import {useSelector, useDispatch} from 'react-redux'
import {getAvailableActions, getPlayerState} from 'logic/game/game-selectors'
import {CARDS} from 'common/cards'
import css from './game-modals.module.scss'
import Button from 'components/button'
import {getGameState} from 'logic/game/game-selectors'
import {PickInfo} from 'common/types/server-requests'

type Props = {
	closeModal: () => void
	info: PickInfo
}
function ChangeHermitModal({closeModal, info}: Props) {
	const dispatch = useDispatch()
	const availableActions = useSelector(getAvailableActions)
	const playerState = useSelector(getPlayerState)
	const gameState = useSelector(getGameState)

	if (info.slot.type !== 'hermit' || !playerState || !gameState || info.rowIndex === undefined) {
		throw new Error('This should never happen')
	}

	const hermitName = info.card?.cardId ? CARDS[info.card.cardId].name : ''
	const hasActiveHermit = playerState.board.activeRow !== null
	const canChange = !hasActiveHermit || availableActions.includes('CHANGE_ACTIVE_HERMIT')

	let message = `Are you sure you want to activate ${hermitName}?`
	if (!canChange) message = `You can not change your active hermit at this moment.`

	const lastAction = hasActiveHermit && canChange

	const handleYes = () => {
		dispatch({type: 'CONFIRM_HERMIT_CHANGE', payload: true})
		closeModal()
	}

	const handleNo = () => {
		dispatch({type: 'CONFIRM_HERMIT_CHANGE', payload: false})
		closeModal()
	}

	return (
		<Modal title="Change active hermit" closeModal={handleNo}>
			<div className={css.confirmModal}>
				<div className={css.description}>
					{lastAction && (
						<div className={css.turnEndNotification}>
							<span className={css.infoIcon}>!</span>
							{/* Hermit change is the last action of your turn. */}
							<p>Changing your hermit will end your turn!</p>
						</div>
					)}
					{message}
				</div>

				<div className={css.options}>
					{canChange ? (
						<>
							<Button onClick={handleNo}>Cancel</Button>
							<Button onClick={handleYes}>Yes</Button>
						</>
					) : (
						<>
							<Button onClick={handleNo}>Ok</Button>
						</>
					)}
				</div>
			</div>
		</Modal>
	)
}

export default ChangeHermitModal
