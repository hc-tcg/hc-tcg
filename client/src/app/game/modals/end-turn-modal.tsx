import {TurnAction} from 'common/types/game-state'
import Button from 'components/button'
import Modal from 'components/modal'
import {getAvailableActions} from 'logic/game/game-selectors'
import {ActionMap} from 'logic/game/tasks/action-modals-saga'
import {getSettings} from 'logic/local-settings/local-settings-selectors'
import {localMessages, useMessageDispatch} from 'logic/messages'
import {useSelector} from 'react-redux'
import css from './game-modals.module.scss'

type Props = {
	closeModal: () => void
}

function EndTurnModal({closeModal}: Props) {
	const availableActions = useSelector(getAvailableActions)
	const dispatch = useMessageDispatch()
	const settings = useSelector(getSettings)

	const handleEndTurn = () => {
		dispatch({type: localMessages.GAME_TURN_END})
		closeModal()
	}

	const handleCancel = () => {
		closeModal()
	}

	if (!settings.confirmationDialogsEnabled) {
		handleEndTurn()
		return null
	}

	let modal = EndTurnModalBody({availableActions, handleCancel, handleEndTurn})

	if (modal !== null) {
		return modal
	}

	handleEndTurn()
}

export function EndTurnModalBody({
	availableActions,
	handleCancel,
	handleEndTurn,
}: {
	availableActions: Array<TurnAction>
	handleCancel?: () => void
	handleEndTurn?: () => void
}) {
	if (availableActions.includes('WAIT_FOR_TURN')) return null

	if (availableActions.every((action) => ActionMap[action] === null)) {
		return null
	}

	return (
		<Modal title="End Turn" closeModal={handleCancel || (() => {})}>
			<div className={css.description}>
				<p>
					Are you sure you want to end your turn? These actions are still
					available:
				</p>
				<hr />
				<ul className={css.availableActions}>
					{availableActions.map((action) => {
						const text = ActionMap[action]
						if (!text) return null
						return (
							<li key={action} className={css.action}>
								{text}
							</li>
						)
					})}
				</ul>
			</div>
			<div className={css.options}>
				<Button variant="default" size="medium" onClick={handleCancel}>
					Cancel
				</Button>
				<Button variant="error" size="medium" onClick={handleEndTurn}>
					End Turn
				</Button>
			</div>
		</Modal>
	)
}

export default EndTurnModal
