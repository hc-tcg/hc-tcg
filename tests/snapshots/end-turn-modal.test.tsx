import {describe, expect, test} from '@jest/globals'
import {EndTurnModalBody} from 'client/app/game/modals/end-turn-modal'
import * as ReactTestRenderer from 'react-test-renderer'

describe('Test End Turn Modal', () => {
	test('No Actions', () => {
		expect(
			ReactTestRenderer.create(
				<EndTurnModalBody availableActions={[]} />,
			).toJSON(),
		).toMatchSnapshot()
	})
	test('All play card actions', () => {
		expect(
			ReactTestRenderer.create(
				<EndTurnModalBody
					availableActions={[
						'PLAY_HERMIT_CARD',
						'PLAY_ITEM_CARD',
						'PLAY_SINGLE_USE_CARD',
						'PLAY_EFFECT_CARD',
						'CHANGE_ACTIVE_HERMIT',
					]}
				/>,
			).toJSON(),
		).toMatchSnapshot()
	})
	test('All attack actions', () => {
		expect(
			ReactTestRenderer.create(
				<EndTurnModalBody
					availableActions={[
						'PRIMARY_ATTACK',
						'SECONDARY_ATTACK',
						'SINGLE_USE_ATTACK',
					]}
				/>,
			).toJSON(),
		).toMatchSnapshot()
	})
	test('All play card actions and attacks', () => {
		expect(
			ReactTestRenderer.create(
				<EndTurnModalBody
					availableActions={[
						'PLAY_HERMIT_CARD',
						'PLAY_ITEM_CARD',
						'PLAY_SINGLE_USE_CARD',
						'PLAY_EFFECT_CARD',
						'CHANGE_ACTIVE_HERMIT',
						'PRIMARY_ATTACK',
						'SECONDARY_ATTACK',
						'SINGLE_USE_ATTACK',
					]}
				/>,
			).toJSON(),
		).toMatchSnapshot()
	})
	test('Waiting actions', () => {
		expect(
			ReactTestRenderer.create(
				<EndTurnModalBody
					availableActions={['WAIT_FOR_TURN', 'WAIT_FOR_OPPONENT_ACTION']}
				/>,
			).toJSON(),
		).toMatchSnapshot()
	})
})
