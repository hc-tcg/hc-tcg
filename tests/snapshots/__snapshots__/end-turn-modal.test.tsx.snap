// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test End Turn Modal All attack actions 1`] = `
[
  <div />,
  <div
    className=""
  >
    <div>
      <span>
        End Turn
      </span>
      <button
        onClick={[Function]}
        tabIndex={-1}
      >
        <img
          alt="close"
          src="/images/CloseX.svg"
        />
      </button>
    </div>
    <div>
       
      <div>
        <p>
          Are you sure you want to end your turn? These actions are still available:
        </p>
        <hr />
        <ul>
          <li>
            Attacking opponent with a primary attack
          </li>
          <li>
            Attacking opponent with a secondary attack
          </li>
          <li>
            Attacking opponent with a single use effect
          </li>
        </ul>
      </div>
      <div>
        <button
          className=""
        >
          Cancel
        </button>
        <button
          className=""
        >
          End Turn
        </button>
      </div>
       
    </div>
  </div>,
]
`;

exports[`Test End Turn Modal All play card actions 1`] = `
[
  <div />,
  <div
    className=""
  >
    <div>
      <span>
        End Turn
      </span>
      <button
        onClick={[Function]}
        tabIndex={-1}
      >
        <img
          alt="close"
          src="/images/CloseX.svg"
        />
      </button>
    </div>
    <div>
       
      <div>
        <p>
          Are you sure you want to end your turn? These actions are still available:
        </p>
        <hr />
        <ul>
          <li>
            Playing a hermit card
          </li>
          <li>
            Playing an item card
          </li>
          <li>
            Playing a single use effect card
          </li>
          <li>
            Playing an attach effect card
          </li>
          <li>
            Changing your active hermit
          </li>
        </ul>
      </div>
      <div>
        <button
          className=""
        >
          Cancel
        </button>
        <button
          className=""
        >
          End Turn
        </button>
      </div>
       
    </div>
  </div>,
]
`;

exports[`Test End Turn Modal All play card actions and attacks 1`] = `
[
  <div />,
  <div
    className=""
  >
    <div>
      <span>
        End Turn
      </span>
      <button
        onClick={[Function]}
        tabIndex={-1}
      >
        <img
          alt="close"
          src="/images/CloseX.svg"
        />
      </button>
    </div>
    <div>
       
      <div>
        <p>
          Are you sure you want to end your turn? These actions are still available:
        </p>
        <hr />
        <ul>
          <li>
            Playing a hermit card
          </li>
          <li>
            Playing an item card
          </li>
          <li>
            Playing a single use effect card
          </li>
          <li>
            Playing an attach effect card
          </li>
          <li>
            Changing your active hermit
          </li>
          <li>
            Attacking opponent with a primary attack
          </li>
          <li>
            Attacking opponent with a secondary attack
          </li>
          <li>
            Attacking opponent with a single use effect
          </li>
        </ul>
      </div>
      <div>
        <button
          className=""
        >
          Cancel
        </button>
        <button
          className=""
        >
          End Turn
        </button>
      </div>
       
    </div>
  </div>,
]
`;

exports[`Test End Turn Modal No Actions 1`] = `null`;

exports[`Test End Turn Modal Waiting actions 1`] = `null`;
